docs/src/literate/topology_optimization.jl

```diff
@@ -0,0 +1,502 @@
+
+mutable struct MaterialState{T, S <: AbstractArray{SymmetricTensor{2, 2, T}, 1}} 
+    χ::T ## density
+    ε::S ## strain in each quadrature point
+end
+
+function MaterialState(ρ, n_qp)
+    return MaterialState(ρ, Array{SymmetricTensor{2,2,Float64},1}(undef, n_qp))
+end
+
+function UpdateMaterialStates!(χn1, states, dh)
+    for (element, state) in zip(CellIterator(dh),states)
+        state.χ = χn1[cellid(element)]
+    end
+end
+#md nothing # hide
+
+# Next, we define a function to calculate the driving forces for all elements.
+# For this purpose, we iterate through all elements and calculate the average strain in each
+# element. Then, we compute the driving force from the formula introduced at the beginning.
+
+function DrivingForces(states, mp, dh)
+    pΨ = zeros(length(states))
+    χn = zeros(length(states))
+    for (element, state) in zip(CellIterator(dh), states)
+        i = cellid(element)
+        ε = sum(state.ε)/length(state.ε) ## average element strain
+        pΨ[i] = 1/2 * mp.p * state.χ^(mp.p-1) * (ε ⊡ mp.C ⊡ ε)
+        χn[i] = state.χ
+    end
+    return pΨ, χn 
+end
+#md nothing # hide
+
+# Now we calculate the Laplacian. For this purpose, we will later create the grid topology of 
+# the grid by using the function `ExclusiveTopology`. Then we iterate through each face of each element,
+# obtaining the neighboring element by using the `getneighborhood` function. For boundary faces,
+# the function call will return an empty object. In that case we use the dictionary to instead find the opposite
+# face, as discussed in the introduction. Then, the approximation of the Laplacian reduces to the sum below. 
+
+function Laplacian(dh, topology, χn, Δh)
+    ∇²χ = zeros(getncells(dh.grid))
+    _nfaces = nfaces(dh.grid.cells[1])
+    opp = Dict(1=>3, 2=>4, 3=>1, 4=>2)
+    nbg = zeros(Int,_nfaces)
+    
+    for element in CellIterator(dh)
+        i = cellid(element)
+        for j in 1:_nfaces
+            nbg_cellid = getcells(getneighborhood(topology, dh.grid, FaceIndex(i,j)))
+            if(!isempty(nbg_cellid))
+                nbg[j] = first(nbg_cellid) ## assuming only one face neighbor per cell
+            else ## boundary face
+                nbg[j] = first(getcells(getneighborhood(topology, dh.grid, FaceIndex(i,opp[j]))))
+            end
+        end
+        
+        ∇²χ[i] = (χn[nbg[1]]+χn[nbg[2]]+χn[nbg[3]]+χn[nbg[4]]-4*χn[i])/(Δh^2)
+    end
+
+    return ∇²χ
+end
+#md nothing # hide
+
+# To ensure that the mass is kept constant, we have to calculate the constraint
+# parameter $\lambda$, which we do via the bisection method. We repeat the calculation
+# until the difference between the current density and target density nearly vanishes. 
+
+function Bisection(χn, Δχ, ρ, ηs, χ_min) 
+    n_el = length(χn)
+    
+    χ_trial = zeros(n_el)
+    ρ_trial = 0.0
+    
+    λ_lower = minimum(Δχ) - ηs
+    λ_upper = maximum(Δχ) + ηs
+    λ_trial = 0.0
+    
+    while(abs(ρ-ρ_trial)>1e-7)
```

Comment:
In general I would agree. In this case, to my knowledge there should not be any instabilities in the mathematical method to prevent the loop from exiting eventually. I can understand, that it might be nice to implement such a feature, but the question remains how to choose a suitable maximum number (the only possibility I see is using a very large number just in case or aborting the program with an error message to prevent the user from obtaining wrong results).

Reply:
You know the algorithm better than I do, so if you say it's not needed, I am fine with that.

Reply:
Sorry to bother again, but can you add a comment here in the example about the convergence guarantee?

Reply:
(See above)